### PR TITLE
Add "deep" flag to heartbeat check, round 2

### PR DESCRIPTION
**Pull request by lloyd**
this is a re-issuing of issue #1800 with unit tests and a small pile of fixes.

This is for issue #1767 - and adds a "deep" heartbeat that is a requirement from ops for train-2012.06.22

### CODE REVIEW

**Comment by lloyd** [comment]
/cc @shane-tomlinson @zaach @fetep @ozten @jrgm 

Can someone review this dude and reply today (wed)?  This is required by @fetep for the train we'll cut on friday

File: lib/heartbeat.js [hunk 1]
```diff
@@ -4,20 +4,74 @@
...
+          } else {
+            logger.warn("heartbeat failed due to dependencies - " + notOk.join(', '));
+            ok(false, '\n' + notOk.join('\n') + '\n');
+          }
+        }
+      };
+    }
+
+    function ok(yeah, msg) {
+      res.writeHead(yeah ? 200 : 500);
```
Comment (ozten) [inline lib/heartbeat.js:63]: Nice, I like the different status code.

File: lib/heartbeat.js [hunk 2]
```diff
@@ -4,20 +4,74 @@
...
+    // callback for checking a dependency
+    function checkCB (num) {
+      return function (err, isOk) {
+        checked++;
+        if (err) {
+          notOk.push(dependencies[num] + ': '+ err);
+        }
+
+        // if all dependencies have been checked
+        if (checked == count) {
```
Comment (ozten) [inline lib/heartbeat.js:45]: This is a little confusing, but I see after a second look why. Looks like a good re-usable way to do this.

File: tests/heartbeat-test.js [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+    });
+  });
+});
+
+// now let's SIGSTOP the browserid process and verify that the router's
+// deep heartbeat fails within 11s
+suite.addBatch({
+  "stopping the browserid process": {
+    topic: function() {
+      process.kill(parseInt(process.env['BROWSERID_PID'], 10), 'SIGSTOP');
```
Comment (ozten) [inline tests/heartbeat-test.js:61]: Wow, neat.

File: tests/heartbeat-test.js [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+    }
+  }
+});
+
+
+start_stop.addShutdownBatches(suite);
+
+// run or export the suite.
+if (process.argv[1] === __filename) suite.run();
+else suite.export(module);
```
Comment (ozten) [inline tests/heartbeat-test.js:122]: Great work here.

**Comment by ozten** [comment]
Unit tests pass http://travis-ci.org/#!/mozilla/browserid/builds/1662853
r+

**Pull request merged**